src/transaction/transaction.rs

```diff
@@ -357,15 +389,56 @@ impl Transaction {
     /// # });
     /// ```
     pub async fn commit(&mut self) -> Result<()> {
-        TwoPhaseCommitter::new(
+        if !matches!(
+            self.status,
+            TransactionStatus::StartedCommit | TransactionStatus::Active
+        ) {
+            return Err(ErrorKind::OperationAfterCommitError.into());
+        }
+        self.status = TransactionStatus::StartedCommit;
+
+        let res = TwoPhaseCommitter::new(
             self.buffer.to_proto_mutations().await,
             self.timestamp.version(),
             self.bg_worker.clone(),
             self.rpc.clone(),
             self.for_update_ts,
         )
         .commit()
-        .await
+        .await;
+
+        if res.is_ok() {
```

Comment:
I'm not sure about the logic here: if the user tries to commit and fails, why shouldn't they be able to modify the transaction before trying again to commit?

Reply:
I think there are 2 reasons:
1. Users seldom need to modify transactions after they attempted to commit or rollback. 
2. It's possible that commit or rollback succeed in the server side but return error in the client side. In such case performing more operations will lead to weird behavior.

Reply:
I think 1 doesn't mean we should prevent it, but 2 is a good reason.